LDP L2VPN move to library

pguibert6WIND opened the pull request:
Extract and Build a new L2VPN library inherited from LDP.
This building block is for the new northbound.
It is also intended for other services which want to do L2VPN.

pguibert6WIND committed:
yang, ldpd: add ldp l2vpn yang data model

The l2vpn entries configured in LDP are a sub-type of the so called
l2vpn components described in ch.3.1 of RFC4664. The series of commits
aims at extracting a data model, first from LDP, then extending the
model, in order to apply the L2VPN model to other instances like BGP.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
ldpd: move l2vpn configuration to yang northbound

Move l2vpn configuration and pw configuration to yang northbound.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
ldpd: move running-config to yang API

Use the yang API to parse the running-config for the l2vpn contexts.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
ldpd: handle dual mgmt/ldp usage of l2vpn

In order for mgmt to support l2vpn configuration, the 'ldp_l2vpn_init()'
function should initialise the cli.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
tests: add l2vpn and pseudowire node to unit tests

Check in unit tests that l2vpn and pseudowire nodes are correctly
handled.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
ldpd: rename local variable to avoid shadow warning

The following compilation warning happens if I include "lde.h" in the
labelmapping.c file.

> ldpd/labelmapping.c: In function ‘recv_labelmessage’:
> ldpd/labelmapping.c:135:34: warning: declaration of ‘ft’ shadows a global declaration [-Wshadow]
>   135 |         struct tlv               ft;
>       |                                  ^~
> In file included from ldpd/ldp_l2vpn.h:18,
>                  from ldpd/ldpe.h:19,
>                  from ldpd/labelmapping.c:12:
> ./ldpd/lde.h:135:26: note: shadowed declaration is here
>   135 | extern struct fec_tree   ft;
>       |                          ^~
> ldpd/labelmapping.c: In function ‘gen_fec_tlv’:
> ldpd/labelmapping.c:529:25: warning: declaration of ‘ft’ shadows a global declaration [-Wshadow]
>   529 |         struct tlv      ft;
>       |                         ^~
> In file included from ldpd/ldp_l2vpn.h:18,
>                  from ldpd/ldpe.h:19,
>                  from ldpd/labelmapping.c:12:
> ./ldpd/lde.h:135:26: note: shadowed declaration is here
>   135 | extern struct fec_tree   ft;
>       |                          ^~
>
It is not very safe to use global variables and local variables with the
same name. Rename local ft variable to avoid confusion.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
ldpd: remove useless functions

Remove useless functions not used.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
ldpd: handle l2vpn hooks upon l2vpn config change

Instead of directly calling the l2vpn code, an hook is called.
- Upon l2vpn or pseudowire changes, hooks are called
The add/delete/change hooks are created.
- A condition is added to NB to check if an iface can be configured in
l2vpn. If LDP is configured on an interface, then L2VPN can not use it.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
ldpd: replace ldpd_addr with g_addr

Use the standard g_addr structure defined in lib instead of the
ldpd_addr.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
ldpd: move l2vpn API to ldp_l2vpn[_cli].[ch] files

The l2vpn structures are isolated from the rest of ldp code.
The main APIs used to handle pw, l2vpn if and l2vpn structure are moved
to ldp_l2vpn_cli.c file.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
ldpd: extract l2vpn_head from ldpd_conf vty structure

The ldp_conf 'vty_conf' global variable contains the l2vpn list
configured by the CLI. Extern the l2vpn_head tree structure
in the ldp_l2vpn library.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
lib: add memory support for l2vpn library

Using malloc avoids needing to control whether the allocation has been
done or not, since the check is builtin.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
ldpd: rename l2vpn.c to ldp_l2vpn.c

The file is being appended the ldp_ prefix. This file contains functions
that are self to ldp.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
ldpd, lib: move l2vpn library to common lib folder

Move and rename library in charge of L2VPN and pseudowire library
support.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
ldpd: replace fatal with assert in l2vpn code

fatal() API is now known in places other than LDPD. In order to be able
to compile, fatal is replaced with assert. This code will use standard
memory API which detect internally if memory allocation failed.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

mjstapp reviewed (commented):
can you say more about how this set of changes would be used? "it's for other services" is a little vague: are you proposing to move bgp L2VPN in some way, for example?

pguibert6WIND committed:
ldpd: rename l2vpn_exit() to ldp_l2vpn_exit()

This function is specific to LDP.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
ldpd: duplicate l2vpn_del() for ldp use cases

L2vpn objects will be handled either in the shared library for common
config, or in LDP code for its own purpose. The memory management of
both objects will be separate. This commit introduces a new ldp_l2vpn_del()
command inherited from l2vpn_del() commmand, and will be used for ldp
purpose only.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
topotests: use unified config in ldp_sync_isis test

Use unified configuration in ldp_sync_isis test.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
topotests: fix holddown timer value applied to all interfaces

With unified configuration, interface are attached to IS-IS, and then
inherit from LDP sync holddown value configured in main IS-IS instance.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

Commit:
topotests: ldp vpls use frr.conf

Signed-off-by: Loïc Sang <loic.sang@6wind.com>

pguibert6WIND committed:
ldpd: fix openbsd compilation error

pfkey.c file
> build	21-Jan-2026 15:26:35	  CC       ldpd/pfkey.o
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:26:18: error: declaration of 'union ldpd_addr' will not be visible outside of this function [-Werror,-Wvisibility]
> build	21-Jan-2026 15:26:35	                    int, union ldpd_addr *, union ldpd_addr *,
> build	21-Jan-2026 15:26:35	                               ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:30:37: error: declaration of 'union ldpd_addr' will not be visible outside of this function [-Werror,-Wvisibility]
> build	21-Jan-2026 15:26:35	static int       pfkey_sa_add(int, union ldpd_addr *, union ldpd_addr *,
> build	21-Jan-2026 15:26:35	                                         ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:32:40: error: declaration of 'union ldpd_addr' will not be visible outside of this function [-Werror,-Wvisibility]
> build	21-Jan-2026 15:26:35	static int       pfkey_sa_remove(int, union ldpd_addr *, union ldpd_addr *,
> build	21-Jan-2026 15:26:35	                                            ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:47:19: error: declaration of 'union ldpd_addr' will not be visible outside of this function [-Werror,-Wvisibility]
> build	21-Jan-2026 15:26:35	    int af, union ldpd_addr *src, union ldpd_addr *dst, uint32_t spi,
> build	21-Jan-2026 15:26:35	                  ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:46:1: error: conflicting types for 'pfkey_send'
> build	21-Jan-2026 15:26:35	pfkey_send(int sd, uint8_t satype, uint8_t mtype, uint8_t dir,
> build	21-Jan-2026 15:26:35	^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:25:13: note: previous declaration is here
> build	21-Jan-2026 15:26:35	static int       pfkey_send(int, uint8_t, uint8_t, uint8_t,
> build	21-Jan-2026 15:26:35	                 ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:69:14: error: incompatible pointer types passing 'union ldpd_addr *' to parameter of type 'const union g_addr *' [-Werror,-Wincompatible-pointer-types]
> build	21-Jan-2026 15:26:35	        addr2sa(af, src, 0, &su_src);
> build	21-Jan-2026 15:26:35	                    ^~~
> build	21-Jan-2026 15:26:35	ldpd/ldpd.h:705:42: note: passing argument to parameter 'addr' here
> build	21-Jan-2026 15:26:35	void addr2sa(int af, const union g_addr *addr, uint16_t port, union sockunion *su);
> build	21-Jan-2026 15:26:35	                                         ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:87:14: error: incompatible pointer types passing 'union ldpd_addr *' to parameter of type 'const union g_addr *' [-Werror,-Wincompatible-pointer-types]
> build	21-Jan-2026 15:26:35	        addr2sa(af, dst, 0, &su_dst);
> build	21-Jan-2026 15:26:35	                    ^~~
> build	21-Jan-2026 15:26:35	ldpd/ldpd.h:705:42: note: passing argument to parameter 'addr' here
> build	21-Jan-2026 15:26:35	void addr2sa(int af, const union g_addr *addr, uint16_t port, union sockunion *su);
> build	21-Jan-2026 15:26:35	                                         ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:339:28: error: declaration of 'union ldpd_addr' will not be visible outside of this function [-Werror,-Wvisibility]
> build	21-Jan-2026 15:26:35	pfkey_sa_add(int af, union ldpd_addr *src, union ldpd_addr *dst, uint8_t keylen,
> build	21-Jan-2026 15:26:35	                           ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:339:1: error: conflicting types for 'pfkey_sa_add'
> build	21-Jan-2026 15:26:35	pfkey_sa_add(int af, union ldpd_addr *src, union ldpd_addr *dst, uint8_t keylen,
> build	21-Jan-2026 15:26:35	^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:30:13: note: previous declaration is here
> build	21-Jan-2026 15:26:35	static int       pfkey_sa_add(int, union ldpd_addr *, union ldpd_addr *,
> build	21-Jan-2026 15:26:35	                 ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:343:10: error: incompatible pointer types passing 'union ldpd_addr *' to parameter of type 'union ldpd_addr *' [-Werror,-Wincompatible-pointer-types]
> build	21-Jan-2026 15:26:35	            af, src, dst, 0, 0, 0, NULL, 0, 0, NULL, 0, 0) < 0)
> build	21-Jan-2026 15:26:35	                ^~~
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:26:29: note: passing argument to parameter here
> build	21-Jan-2026 15:26:35	                    int, union ldpd_addr *, union ldpd_addr *,
> build	21-Jan-2026 15:26:35	                                          ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:343:15: error: incompatible pointer types passing 'union ldpd_addr *' to parameter of type 'union ldpd_addr *' [-Werror,-Wincompatible-pointer-types]
> build	21-Jan-2026 15:26:35	            af, src, dst, 0, 0, 0, NULL, 0, 0, NULL, 0, 0) < 0)
> build	21-Jan-2026 15:26:35	                     ^~~
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:26:48: note: passing argument to parameter here
> build	21-Jan-2026 15:26:35	                    int, union ldpd_addr *, union ldpd_addr *,
> build	21-Jan-2026 15:26:35	                                                             ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:348:10: error: incompatible pointer types passing 'union ldpd_addr *' to parameter of type 'union ldpd_addr *' [-Werror,-Wincompatible-pointer-types]
> build	21-Jan-2026 15:26:35	            af, src, dst, *spi, 0, keylen, key, 0, 0, NULL, 0, 0) < 0)
> build	21-Jan-2026 15:26:35	                ^~~
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:26:29: note: passing argument to parameter here
> build	21-Jan-2026 15:26:35	                    int, union ldpd_addr *, union ldpd_addr *,
> build	21-Jan-2026 15:26:35	                                          ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:348:15: error: incompatible pointer types passing 'union ldpd_addr *' to parameter of type 'union ldpd_addr *' [-Werror,-Wincompatible-pointer-types]
> build	21-Jan-2026 15:26:35	            af, src, dst, *spi, 0, keylen, key, 0, 0, NULL, 0, 0) < 0)
> build	21-Jan-2026 15:26:35	                     ^~~
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:26:48: note: passing argument to parameter here
> build	21-Jan-2026 15:26:35	                    int, union ldpd_addr *, union ldpd_addr *,
> build	21-Jan-2026 15:26:35	                                                             ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:356:31: error: declaration of 'union ldpd_addr' will not be visible outside of this function [-Werror,-Wvisibility]
> build	21-Jan-2026 15:26:35	pfkey_sa_remove(int af, union ldpd_addr *src, union ldpd_addr *dst,
> build	21-Jan-2026 15:26:35	                              ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:356:1: error: conflicting types for 'pfkey_sa_remove'
> build	21-Jan-2026 15:26:35	pfkey_sa_remove(int af, union ldpd_addr *src, union ldpd_addr *dst,
> build	21-Jan-2026 15:26:35	^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:32:13: note: previous declaration is here
> build	21-Jan-2026 15:26:35	static int       pfkey_sa_remove(int, union ldpd_addr *, union ldpd_addr *,
> build	21-Jan-2026 15:26:35	                 ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:360:10: error: incompatible pointer types passing 'union ldpd_addr *' to parameter of type 'union ldpd_addr *' [-Werror,-Wincompatible-pointer-types]
> build	21-Jan-2026 15:26:35	            af, src, dst, *spi, 0, 0, NULL, 0, 0, NULL, 0, 0) < 0)
> build	21-Jan-2026 15:26:35	                ^~~
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:26:29: note: passing argument to parameter here
> build	21-Jan-2026 15:26:35	                    int, union ldpd_addr *, union ldpd_addr *,
> build	21-Jan-2026 15:26:35	                                          ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:360:15: error: incompatible pointer types passing 'union ldpd_addr *' to parameter of type 'union ldpd_addr *' [-Werror,-Wincompatible-pointer-types]
> build	21-Jan-2026 15:26:35	            af, src, dst, *spi, 0, 0, NULL, 0, 0, NULL, 0, 0) < 0)
> build	21-Jan-2026 15:26:35	                     ^~~
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:26:48: note: passing argument to parameter here
> build	21-Jan-2026 15:26:35	                    int, union ldpd_addr *, union ldpd_addr *,
> build	21-Jan-2026 15:26:35	                                                             ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:374:29: error: incompatible pointer types passing 'union g_addr *' to parameter of type 'union ldpd_addr *' [-Werror,-Wincompatible-pointer-types]
> build	21-Jan-2026 15:26:35	                if (pfkey_sa_add(nbr->af, &nbr->laddr, &nbr->raddr,
> build	21-Jan-2026 15:26:35	                                          ^~~~~~~~~~~
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:30:48: note: passing argument to parameter here
> build	21-Jan-2026 15:26:35	static int       pfkey_sa_add(int, union ldpd_addr *, union ldpd_addr *,
> build	21-Jan-2026 15:26:35	                                                    ^
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:374:42: error: incompatible pointer types passing 'union g_addr *' to parameter of type 'union ldpd_addr *' [-Werror,-Wincompatible-pointer-types]
> build	21-Jan-2026 15:26:35	                if (pfkey_sa_add(nbr->af, &nbr->laddr, &nbr->raddr,
> build	21-Jan-2026 15:26:35	                                                       ^~~~~~~~~~~
> build	21-Jan-2026 15:26:35	ldpd/pfkey.c:30:67: note: passing argument to parameter here
> build	21-Jan-2026 15:26:35	static int       pfkey_sa_add(int, union ldpd_addr *, union ldpd_addr *,
> build	21-Jan-2026 15:26:35	                                                                       ^
> build	21-Jan-2026 15:26:35	fatal error: too many errors emitted, stopping now [-ferror-limit=]
> build	21-Jan-2026 15:26:35	20 errors generated.
> build	21-Jan-2026 15:26:35	gmake[1]: *** [Makefile:11031: ldpd/pfkey.o] Error 1
>

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>
